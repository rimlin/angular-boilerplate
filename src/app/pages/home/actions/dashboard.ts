import { Action } from '@ngrx/store';
import { IItem } from '../../../shared/interfaces/item.interface';
import { ItemModel } from '../../../shared/models';

export const ADD_ITEM = '[Dashboard] Add Item';
export const LOAD_ITEM = '[Dashboard] Load Item';
export const TOGGLE_SELECT = '[Dashboard] Select';
export const TOGGLE_SELECT_SUCCESS = '[Dashboard] Select Success';

export class AddItem implements Action {
  readonly type = ADD_ITEM;

  constructor(public payload: IItem) {}
}

export class LoadItem implements Action {
  readonly type = LOAD_ITEM;

  constructor(public payload: number) {}
}

export class ToggleSelect implements Action {
  readonly type = TOGGLE_SELECT;

  constructor(public payload: number) {}
}

export class ToggleSelectSuccess implements Action {
  readonly type = TOGGLE_SELECT_SUCCESS;

  constructor(public payload: ItemModel) {}
}

export type Actions = AddItem | LoadItem | ToggleSelect | ToggleSelectSuccess;
