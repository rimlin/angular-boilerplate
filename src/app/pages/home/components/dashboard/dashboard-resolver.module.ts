import { Injectable } from '@angular/core';
import { Resolve, ActivatedRouteSnapshot, RouterStateSnapshot } from '@angular/router';
import { Observable } from 'rxjs/Observable';

@Injectable()
export class DashboardResolver implements Resolve<any> {
    resolve(
        route: ActivatedRouteSnapshot,
        state: RouterStateSnapshot
    ): Observable<any> | Promise<any> | any {
        return true;
    }
}
