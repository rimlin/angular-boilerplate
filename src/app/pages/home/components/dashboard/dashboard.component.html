
<br/>
<button (click)="addItem()">Add</button>

<div *ngFor="let item of items">
  <wf-dashboard-item-container [itemModel]="item"></wf-dashboard-item-container>
</div>

