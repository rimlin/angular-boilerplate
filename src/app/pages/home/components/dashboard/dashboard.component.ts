import { Component, ChangeDetectionStrategy, OnInit, Input } from '@angular/core';

@Component({
    selector: 'wf-dashboard',
    changeDetection: ChangeDetectionStrategy.OnPush,
    templateUrl: './dashboard.component.html',
    styleUrls: ['./dashboard.component.scss']
})
export class DashboardComponent implements OnInit {
    @Input() items;

    constructor() { }

    ngOnInit() { }
}
