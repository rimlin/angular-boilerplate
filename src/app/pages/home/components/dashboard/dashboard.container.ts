import 'rxjs/add/operator/let';
import { Component, ChangeDetectionStrategy, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';

import * as fromHome from '../../reducers';
import { ItemModel } from '../../../../shared/models';

@Component({
  selector: 'wf-dashboard-container',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    Items: 
    <wf-dashboard [items]="items$ | async"></wf-dashboard>
  `
})
export class DashboardContainer implements OnInit {
  items$: Observable<ItemModel[]>;

  constructor(private store: Store<fromHome.State>) {
    this.items$ = store.select(fromHome.getAllDashboardItems);
  }

  ngOnInit() {
      
  }
}
