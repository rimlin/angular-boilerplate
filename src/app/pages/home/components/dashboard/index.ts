export * from './dashboard.container';
export * from './dashboard-resolver.module';