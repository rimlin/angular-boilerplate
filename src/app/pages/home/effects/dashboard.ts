import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/skip';
import 'rxjs/add/operator/takeUntil';
import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { Scheduler } from 'rxjs/Scheduler';
import { async } from 'rxjs/scheduler/async';
import { empty } from 'rxjs/observable/empty';
import { of } from 'rxjs/observable/of';

import * as dashboard from '../actions/dashboard';
import { ItemModel } from '../../../shared/models';
import { IItem } from '../../../shared/interfaces/item.interface';

@Injectable()
export class DashboardEffects {
  @Effect()
  addItem$: Observable<Action> = this.actions$
    .ofType<dashboard.LoadItem>(dashboard.LOAD_ITEM)
    .map(action => action.payload)
    .switchMap(itemId => {
      return Observable.of(new dashboard.AddItem({ id: itemId, name: `Effect ${itemId}` }));
    });

  constructor(
    private actions$: Actions,
  ) {}
}
