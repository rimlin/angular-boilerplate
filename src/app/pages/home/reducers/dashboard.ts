import { createSelector } from '@ngrx/store';
import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';

import { ItemModel } from '../../../shared/models';
import { IItem } from '../../../shared/interfaces/item.interface';
import * as dashboard from '../actions/dashboard';

export interface State extends EntityState<ItemModel> {
  selectedItemId: number | null;
}

export const adapter: EntityAdapter<ItemModel> = createEntityAdapter<ItemModel>({
  selectId: (item: ItemModel) => item.id,
  sortComparer: false,
});

export const initialState: State = adapter.getInitialState({
  selectedItemId: null,
});

export function reducer(
  state = initialState,
  action: dashboard.Actions
): State {
  switch (action.type) {
    case dashboard.ADD_ITEM: {
      return {
        ...adapter.addOne(new ItemModel(action.payload), state),
        selectedItemId: state.selectedItemId,
      };
    }

    case dashboard.TOGGLE_SELECT_SUCCESS: {
      return {
        ...adapter.updateOne({ id: action.payload.id, changes: action.payload }, state),
        selectedItemId: action.payload.selected ? action.payload.id : state.selectedItemId,
      };
    }

    default: {
      return state;
    }
  }
}

export const getSelectedId = (state: State) => state.selectedItemId;
