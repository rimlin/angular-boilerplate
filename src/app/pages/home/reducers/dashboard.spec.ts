import { reducer } from './dashboard';
import * as fromDashboard from './dashboard';
import * as dashboard from '../actions/dashboard';
import { ItemModel } from '../../../shared/models';

describe('Dashboard reducer', () => {
  const item = new ItemModel({ id: 0, name: 'Test item' });

  const initialState: fromDashboard.State = {
    ids: [item.id],
    entities: {
      [item.id]: item,
    },
    selectedItemId: null,
  }

  describe('ADD_ITEM', () => {
    test('should add item to initial state', () => {
      const action = new dashboard.AddItem({
        id: 1,
        name: 'Test',
      });

      const result = reducer(fromDashboard.initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('TOGGLE_SELECT_SUCCESS', () => {
    test('should select new item id', () => {
      item.selected = true;

      const action = new dashboard.ToggleSelectSuccess(item);

      const result = reducer(initialState, action);

      expect(result.selectedItemId).toBe(item.id);
    });

    test('selected attribute should be true in selected item', () => {
      item.selected = true;

      const action = new dashboard.ToggleSelectSuccess(item);

      const result = reducer(initialState, action);

      expect(result.entities[0].selected).toBe(true);
    });
  });

  describe('Selectors', () => {
    describe('getSelectedId', () => {
      it('should return the selected id', () => {
        const result = fromDashboard.getSelectedId({
          ...initialState,
          selectedItemId: item.id,
        });

        expect(result).toMatchSnapshot();
      });
    });
  });
});
