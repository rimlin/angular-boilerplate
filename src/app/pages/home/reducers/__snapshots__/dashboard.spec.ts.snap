// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard reducer ADD_ITEM should add item to initial state 1`] = `
Object {
  "entities": Object {
    "1": ItemModel {
      "id": 1,
      "name": "Test",
      "selected": undefined,
    },
  },
  "ids": Array [
    1,
  ],
  "selectedItemId": null,
}
`;

exports[`Dashboard reducer Selectors getSelectedId should return the selected id 1`] = `0`;
